doc/reading/tma-modeling-in-depth.md

```diff
@@ -168,6 +176,298 @@ Another example is the following Figure 4:
 In this figure, `I2`, `I3` and `I10` are the only IterDomains that could run out of boundary,
 they are all TMA-protected. So we can achieve strong correctness.
 
-TODO: explain how to upgrade weak correctness to strong correctness by initializing and predicating?
+## The unachievability of strong correctness for indivisible element stride
+
+Let's take a look at the example in the following Figure 5:
+
+![Figure 5: Indivisible strided tile](tma-modeling-in-depth/no-strong-correctness-schedule.svg)
+
+TMA-protected IterDomains are `I1`, `I2`, and `I3`.
+We can see that `I4` is indivisibly split but not TMA-protected,
+so there exist out-of-boundary index of `I4` that is not eventually translated to out-of-boundary index of `I1`.
+For example, let $i_3$, $i_4$, and $i_1$ be the indices of `I3`, `I4`, and `I1`,
+if $i_3 = 0$ and $i_4 = 4$, then $i_1 = 4$ is a hole but in-boundary for global memory.
+To make it even worse, the index of `I4` depends on the index of `I5`,
+which is hardware parallelized and the programmer has no access to,
+so it is not even possible to write a `if` check for the boundary of `I4`.
+For this example, the data loaded in shared memory looks like the following Figure 6:
+
+![Figure 6: Indivisible strided tile data](tma-modeling-in-depth/no-strong-correctness-data.svg)
+
+From the above figure, we see that there are two tiles where the first half tile is blue and the second half tile is red.
+However, the granularity on TMA that programmer has control of is per tile.
+There is no way for a programmer to tell the hardware that they want part of a tile to be filled with data but part of it filled with zero even if they are not out of the boundary of global memory.
+
+The above observation leads to the following theorem:
+
+**Theorem 6:** It is impossible to achieve strong correctness if and only if there exist a dimension that satisfies all of the following conditions:
```

Comment:
nit: perhaps easier to remember to rephrase this as:
 "The strong correctness is achieved only if the following conditions are satisfied."

Reply:
This is not what Theorem 6 is saying. What Theorem 6 says is, if the condition is satisfied, then no matter what you do, you can not achieve strong correctness. If the condition is not satisfied, then you can achieve strong correctness by doing something.

So there are three levels:
- If there is no indivisible splits on TMA-unprotected IDs, then strong correctness is automatically achieved without needing to do anything
- If there are indivisible splits on TMA-unprotected IDs, but the condition of FTTC is not satisfied, then strong correctness is achievable by predicating these TMA-unprotected IDs.
- If the condition of FTTC is satisfied, then you have no choice but to raise an error.

Reply:
What I wanted to suggest was also to change the conditions because it sounded like a double negative. For example, the following levels would be phrased like:

> but the condition of FTTC is satisfied, then strong correctness is achievable by predicating these TMA-unprotected IDs.

> If the condition of FTTC is not satisfied, then you have no choice but to raise an error.

Just a nit suggestion

Reply:
I am not sure, but if rephrased like "Strong correctness is achievable if no dim satisfies both of the following conditions: 1, 2", then when we say "the condition of theorem 6", it is hard to tell if it refers to "no dim satisfies 1 and 2" or refers to "1, 2".